interfaces/builtin/dbus.go

```diff
@@ -0,0 +1,381 @@
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	// convert name to AppArmor dbus interface (eg, 'org.foo' to 'org.foo{,.*}')
+	var ifaceBuf bytes.Buffer
+	ifaceBuf.WriteString(`"`)
+	ifaceBuf.WriteString(name)
+	ifaceBuf.WriteString(`{,.*}"`)
+
+	old = []byte("###DBUS_INTERFACE###")
+	new = ifaceBuf.Bytes()
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	return snippet
+}
+
+func (iface *DbusInterface) PermanentPlugSnippet(plug *interfaces.Plug, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	return nil, nil
+}
+
+func (iface *DbusInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	bus, name, err := iface.getAttribs(plug.Attrs)
+	if err != nil {
+		return nil, err
+	}
+
+	busSlot, nameSlot, err := iface.getAttribs(slot.Attrs)
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure that we only connect to slot with matching attributes
+	if bus != busSlot || name != nameSlot {
+		return nil, nil
+	}
+
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		// well-known DBus name-specific connected plug policy
+		snippet := getAppArmorSnippet([]byte(dbusConnectedPlugAppArmor), bus, name)
+
+		old := []byte("###SLOT_SECURITY_TAGS###")
+		new := slotAppLabelExpr(slot)
+		snippet = bytes.Replace(snippet, old, new, -1)
+
+		return snippet, nil
+	case interfaces.SecuritySecComp:
+		return []byte(dbusConnectedPlugSecComp), nil
+	}
+	return nil, nil
+}
+
+func (iface *DbusInterface) PermanentSlotSnippet(slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	switch securitySystem {
+	case interfaces.SecurityAppArmor:
+		bus, name, err := iface.getAttribs(slot.Attrs)
+		if err != nil {
+			return nil, err
+		}
+
+		snippets := bytes.NewBufferString("")
+
+		// well-known DBus name-specific permanent slot policy
+		snippet := getAppArmorSnippet([]byte(dbusPermanentSlotAppArmor), bus, name)
+
+		// abstraction policy
+		abstraction, err := getAppArmorAbstraction(bus)
+		if err != nil {
+			return nil, err
+		}
+
+		old := []byte("###DBUS_ABSTRACTION###")
+		new := []byte(abstraction)
+		snippet = bytes.Replace(snippet, old, new, -1)
+
+		snippets.Write(snippet)
+
+		if release.OnClassic {
+			// classic-only policy
+			snippets.Write(getAppArmorSnippet([]byte(dbusPermanentSlotAppArmorClassic), bus, name))
+		}
+
+		return snippets.Bytes(), nil
+	case interfaces.SecuritySecComp:
+		return []byte(dbusPermanentSlotSecComp), nil
+	}
+	return nil, nil
+}
+
+func (iface *DbusInterface) ConnectedSlotSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
+	bus, name, err := iface.getAttribs(slot.Attrs)
+	if err != nil {
+		return nil, err
+	}
+
+	busPlug, namePlug, err := iface.getAttribs(plug.Attrs)
+	if err != nil {
+		return nil, err
+	}
+
+	// ensure that we only connect to slot with matching attributes
+	if bus != busPlug || name != namePlug {
```

Comment:
I haven't looked at this terribly closely, but it looks like this is cause by ResolveConnect() in interfaces/repo.go:

```
func (r *Repository) ResolveConnect(plugSnapName, plugName, slotSnapName, slotName string) (ConnRef, error) {
...
        if slotName == "" {                                                     
                // Find the unambiguous slot that satisfies plug requirements   
                var candidates []string                                         
                for candidateSlotName, candidateSlot := range r.slots[slotSnapName] {
                        // TODO: use some smarter matching (e.g. against $attrs)
                        if candidateSlot.Interface == plug.Interface {          
                                candidates = append(candidates, candidateSlotName)
                        }                                                       
                }
```

Note `// TODO: use some smarter matching (e.g. against $attrs)` - it isn't matching against attrs so the interface code must when matching like this. This makes sense with my observations that I could not error and instead must silently ignore attribute mismatches since Connected*Snippet  was getting each combination.